Don’t print module selectors in more cases + Disable preserve-types when using module selectors

beccadax opened the pull request:
There were a number of cases where Swift would emit module selectors into module interface files that it was subsequently unable to parse because the compiler at least temporarily represented them as dependent member types. Modify the carve-out so that if *any* of a type’s parents are generic parameters or archetypes, we don’t print a module selector on it.

Fixes rdar://166180424.

The old `-module-interface-preserve-types-as-written` workaround flag prevents module selectors from being printed into module interfaces even when they have been explicitly requested. Disable it and emit a warning when it’s used in combination with `-enable-module-selectors-in-module-interface`.

Fixes rdar://166237384.

(These have been rolled into a single PR because the test changes intersect.)

beccadax committed:
Don’t print module selectors in more cases

There were a number of cases where Swift would emit module selectors into module interface files that it was subsequently unable to parse because the compiler at least temporarily represented them as dependent member types. Modify the carve-out so that if *any* of a type’s parents are generic parameters or archetypes, we don’t print a module selector on it.

Fixes rdar://166180424.

beccadax commented:
@swift-ci please smoke test

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -6143,7 +6143,10 @@ class TypePrinter : public TypeVisitor<TypePrinter, void, NonRecursivePrintOptio
       parent = generic->getParent();
-    return parent && parent->isTypeParameter();
+    else if (auto dmt = T->getAs<DependentMemberType>())
+      parent = dmt->getBase();
```
The base of a DependentMemberType should always be a type parameter, so you can just return `true` here without recursing

beccadax committed:
Disable preserve-types when using module selectors

The old `-module-interface-preserve-types-as-written` workaround flag prevents module selectors from being printed into module interfaces even when they have been explicitly requested. Disable it and emit a warning when it’s used in combination with `-enable-module-selectors-in-module-interface`.

Fixes rdar://166237384.
